### PR TITLE
[Infer] Colossal-Inference serving example w/ TorchServe (single GPU case)

### PR DESCRIPTION
**Pull request by yuanheng-zhao**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 📝 What does this PR do?

- This PR adds necessary files for Colossal Inference serving with TorchServe. Steps of deployment were added into README.
- This demo is used for testing and demonstrating the usage of Colossal Inference from `colossalai.inference.tensor_parallel` with TorchServe. For now, serving for single-gpu inference is supported.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Review by tiandiao123 — APPROVED** [review]

File: examples/inference/serving/torch_serve/Colossal_Inference_Handler.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+            self.tokenizer = AutoTokenizer.from_pretrained(model_dir_path, return_tensors="pt")
+        else:
+            logger.warning(f"Model type {inference_config['model_type']} not supported yet.")
+
+        logger.info("Transformer model from path %s loaded successfully", model_dir)
+
+        # NOTE world_size, rank, host, port here are used to launch colossalai dist environment
+        # This world_size is different from the world size of TorchServe
+        world_size = int(os.getenv("WORLD_SIZE", self.tp_size))
+        assert world_size == 1, "Colossal-Inference with tensor parallel is not supported on TorchServe for now"
```
[the remark this answers is not in the document]
Comment (yuanheng-zhao) [reply examples/inference/serving/torch_serve/Colossal_Inference_Handler.py:105]: Single-gpu inference serving is supported for now (mentioned in readme). Colossalai tensor parallel will be supported in later version.

File: examples/inference/serving/torch_serve/Colossal_Inference_Handler.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        gpu_id = ctx.system_properties.get("gpu_id", -1)
+        model_dir = ctx.system_properties.get("model_dir")
+
+        # Inference configs are collected together in model yaml config for handler use
+        inference_config = ctx.model_yaml_config["handler"]
+        logger.info(inference_config)
+        inference_config["model_type"]
+        self.tp_size = inference_config.get("tp_size", 1)
+        self.max_batch_size = inference_config.get("max_batch_size", 4)
+        self.max_input_len = inference_config.get("max_input_len", 1024)
```
[the remark this answers is not in the document]
Comment (yuanheng-zhao) [reply examples/inference/serving/torch_serve/Colossal_Inference_Handler.py:75]: These are default values if the properties are not assigned in config file

File: examples/inference/serving/torch_serve/Colossal_Inference_Handler.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+from colossalai.shardformer import ShardConfig
+
+logger = logging.getLogger(__name__)
+logger.info("Transformers version %s", transformers.__version__)
+logger.info("ColossalAI version %s", colossalai.__version__)
+
+
+# from colossalai.testing
+# assins a random port, for demo use only
+def free_port() -> int:
```
Comment (ver217) [inline examples/inference/serving/torch_serve/Colossal_Inference_Handler.py:24]: why not import from colossalai directly?

File: examples/inference/serving/torch_serve/Colossal_Inference_Handler.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+        """
+        input_ids_batch, attention_mask_batch = input_batch
+        inferences = []
+
+        # mode: text_generation
+        input_ids_batch = input_ids_batch.to(self.device)
+        outputs = self.infer_engine.generate(
+            dict(input_ids=input_ids_batch, attention_mask=attention_mask_batch),
+            do_sample=True,
+            top_p=0.95,
```
Comment (ver217) [inline examples/inference/serving/torch_serve/Colossal_Inference_Handler.py:189]: These args should be configurable

File: examples/inference/serving/torch_serve/Colossal_Inference_Handler.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        self.max_input_len = inference_config.get("max_input_len", 1024)
+        self.max_output_len = inference_config.get("max_output_len", 128)
+
+        self.device = torch.device("cuda:" + str(gpu_id) if torch.cuda.is_available() and gpu_id >= 0 else "cpu")
+        logger.info(f"Device set to {self.device}")
+        logger.info(f"torch.cuda.device_count() {torch.cuda.device_count()}")
+
+        logger.info(f"Unpacking from model_dir {model_dir}")
+        model_dir_path = os.path.join(model_dir, "model")
+        with zipfile.ZipFile(model_dir + "/model.zip", "r") as zip_ref:
```
Comment (ver217) [inline examples/inference/serving/torch_serve/Colossal_Inference_Handler.py:84]: What's the format of model?

Comment (yuanheng-zhao) [reply examples/inference/serving/torch_serve/Colossal_Inference_Handler.py:84]: A regular model repo or a snapshot on HuggingFace, which contains files like pytorch_modelXX.bin, config.json, tokenizer.json, etc

File: examples/inference/serving/torch_serve/Colossal_Inference_Handler.py (hunk 4 above)
Comment (yuanheng-zhao) [reply examples/inference/serving/torch_serve/Colossal_Inference_Handler.py]: Revised. Now these args could be assigned in yaml config file, or just default values if not assigned in config

File: examples/inference/serving/torch_serve/Colossal_Inference_Handler.py (hunk 3 above)
Comment (yuanheng-zhao) [reply examples/inference/serving/torch_serve/Colossal_Inference_Handler.py]: Revised. Now imports from colossalai.testing

**Review by FoolPlayer — APPROVED** [review]

**Review by FrankLeeeee — APPROVED** [review]

**Pull request merged**